WIP: Allow wallet key import RPCs to track TxOut amounts on -prune nodes (on top of #9306)

ryanofsky opened the pull request:
Allow importprivkey, importaddress, importpubkey, and importmulti RPCs to be
called with rescan=True option on nodes with pruned blockchains (started with
-prune option). Instead of returning errors, the RPCs will now look for
transactions spending to the key in the UTXO database, and display them in the
wallet.

Because the UTXO database doesn't store full transaction information (it
discards TxIns) this change introduces a new "fIncomplete" CWalletTx member to
distinguish these from normal wallet transactions.

This change makes the -prune mode usable with wallets that need to import keys
after the initial sync.

Fixes #8497.

ryanofsky committed:
Allow wallet key import RPCs to track TxOut amounts on -prune nodes

Allow importprivkey, importaddress, importpubkey, and importmulti RPCs to be
called with rescan=True option on nodes with pruned blockchains (started with
-prune option). Instead of returning errors, the RPCs will now look for
transactions spending to the key in the UTXO database, and display them in the
wallet.

Because the UTXO database doesn't store full transaction information (it
discards TxIns) this change introduces a new "fIncomplete" CWalletTx member to
distinguish these from normal wallet transactions.

This change makes the -prune mode usable with wallets that need to import keys
after the initial sync.

Fixes #8497.

jonasschnelli commented on src/primitives/transaction.cpp:
```diff
@@ -62,9 +62,14 @@ uint256 CMutableTransaction::GetHash() const
+void CTransaction::SetHash(const uint256 &tx_hash)
 {
-    *const_cast<uint256*>(&hash) = SerializeHash(*this, SER_GETHASH, SERIALIZE_TRANSACTION_NO_WITNESS);
+    *const_cast<uint256*>(&hash) = tx_hash;
```
Hmm... this would result in not longer coupling the hash with the data-structure. This can result in difficult states for future changes.
Would a new primitive (maybe `CIncompleteTransaction` or similar) make more sense?

ryanofsky replied:
I don't know if @sipa's suggestion of making a new `CTransaction` constructor that accepts a hash would sufficiently address your concern. (It would not eliminate the risk that some code could make a seemingly harmless `UpdateHash()` call on the wallet transaction and overwrite the correct hash).

Personally, the way I would go about this would be to change `CWalletTx` not to inherit from `CMerkleTx`, and instead just contain either a `variant<CMerkleTx, CCoins>` or an  `optional<CMerkleTx>` member. I can try this in a separate PR (since would create a lot of unrelated code churn in the wallet) if you are amenable to the idea.

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -1506,6 +1507,29 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
 
+void CWallet::ScanForWalletUTXOs()
+{
+    for (std::unique_ptr<CCoinsViewCursor> pcursor(pcoinsTip->Cursor());
```
Would there be a better way to keep the wallet separated from the UTXO set?
During wallet creation (`init.cpp`) it could pass-in a lambda that would take care about finding the corresponding `CCoins`.

luke-jr replied:
Won't this miss unconfirmed TXOs?

ryanofsky replied:
This doesn't deal with unconfirmed TXOs, but neither does `ScanForWalletTransactions`, which this is just intended to be a lightweight fallback for.

I don't know the wallet code well enough to know how it does deal with unconfirmed TXOs, but I expect that if there are issues tracking them, they would be orthogonal to this change.

jonasschnelli commented:
Conceptual questions:
* I guess the new `importmulti` call should also be covered in this change?
* Idea: We could deprecate the other import calls in favor of `importmulti` (which has more features and makes generally more sense).

* Does spending those incomplete transactions work well? IMO some checks (isFinal()) do read the nSequence value of the (now missing) inputs?

Things that would need to be done:
-> Visual attribution in the GUI
-> the `fIncomplete` status should probably be added to the RPC TxToJSON (or similar) function
-> Add tests

sipa commented:
NACK when done this way. If you need an incomplete transaction, use CMutableTransaction instead of hacking that functionality around CTransaction.

luke-jr commented on src/wallet/rpcdump.cpp:
```diff
@@ -398,7 +395,10 @@ UniValue importpubkey(const JSONRPCRequest& request)
+        if (fPruneMode)
+          pwalletMain->ScanForWalletUTXOs();
+        else
+          pwalletMain->ScanForWalletTransactions(chainActive.Genesis(), true);
```
Shouldn't there be some way to scan for UTXOs on non-pruned nodes?

ryanofsky replied:
I could definitely add it as another option to the RPC. I guess the benefit to the RPC caller would be that if they're importing a key which they already know is unspent, calling `ScanForWalletUTXOs` would be faster than calling `ScanForWalletTransactions`?

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -1506,6 +1507,29 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
+        if (!pcursor->GetKey(hash))
+            LogPrintf("GetKey failed");
+        else if (!pcursor->GetValue(coins))
+            LogPrintf("GetValue failed");
```
These errors leave much to be desired...

ryanofsky commented:
Thanks @jonasschnelli, I am planning to get to the various todos you mentioned. The one I might like to know more about is "Visual attribution in the GUI." Do you have any thoughts on what this should look like? And what would be the reasons for making the distinction? E.g. are there particular wallet features or RPC calls that won't work with incomplete transaction data? (Asking because I actually haven't used the wallet UI very much myself yet.)

jonasschnelli commented:
@ryanofsky: I think the GUI stuff can be done in a separate PR after this has been merged.

ryanofsky commented:
Thanks @jonasschnelli, that makes sense. I still would appreciate knowing any thoughts you may have on the GUI implementation either here or in the original issue #8497.

sipa commented:
I'm working on a series of changes to the serialization code, the result of which will be that CTransaction is fully immutable, so UpdateHash will go away.

ryanofsky commented:
This is better tested now and ready to be reviewed and merged. There are a few extra things that could be done based on comments above, but they don't necessarily have to be part of this PR:

- [ ] Maybe add test to confirm IsFinal works after an incomplete transaction output is spent. ([jonasschnelli comment](https://github.com/bitcoin/bitcoin/pull/9137#issuecomment-260110132))
- [ ] Maybe add visual attribution for incomplete transactions in gui ([jonasschnelli comment](https://github.com/bitcoin/bitcoin/pull/9137#issuecomment-260110132))
- [ ] Maybe add incomplete status to TxToJSON ([jonasschnelli comment](https://github.com/bitcoin/bitcoin/pull/9137#issuecomment-260110132))
- [ ] Maybe allow UTXO rescanning on non -prune nodes ([luke-jr comment](https://github.com/bitcoin/bitcoin/pull/9137#discussion_r87690967))
- [ ] Maybe add ability for wallet to scan mempool for when importing keys ([luke-jr comment](https://github.com/bitcoin/bitcoin/pull/9137#pullrequestreview-8321654))

ryanofsky commented:
Added test code to `import-rescan.py` to make sure spends of incomplete transactions are picked up by the wallet and finalized.

ryanofsky commented:
Updated to fix conflicts after #8580 merge.

ryanofsky committed:
Make CCoinsViewCache::Cursor() return latest data

Change CCoinsViewCache::Cursor() method to return a cursor yielding the latest
CCoins entries, instead of just previous entries prior to the last cache flush.

The CCoinsViewCache::Cursor method is not currently used. This change just
enables new features that rely on scanning the UXTO set to work correctly (for
example https://github.com/bitcoin/bitcoin/pull/9152, which adds a
sweepprivkeys RPC, and https://github.com/bitcoin/bitcoin/pull/9137, which
improves handling of imported keys for nodes with pruning enabled.)

gmaxwell commented:
rescan=true and importing pruned things are not the same though. Rescan makes sure you get the historical transactions and a utxo scan doesn't.   I think we may want to track a global state for a wallet that indicates if it believes its history is complete or not, and report that.  (e.g. in the GUI there could be a bar at the top of the transaction list "some historical transactions may be missing")

ryanofsky commented:
Having a gui indication of an incomplete historical transaction list seems like a good idea with this change, but it would probably also be a good idea without this change. (The transaction list will be incomplete now if you import with a key without rescanning, or if you use the removeprunedfunds API.) The goal of this change is to make possible to see and spend balances after importing keys on `-prune` nodes as described in #8497.

ryanofsky commented:
I do think it would be a good idea not overload to the meaning of `rescan=true`, though. Maybe `rescan` could be changed from a bool to a string argument where `rescan="history"` would scan the blockchain and `rescan="utxo"` would scan the utxo set. `rescan=true` could still be kept for backwards compatibility.

ryanofsky commented:
Haven't worked on this in a while. Will close and reopen if I make more more progress.

sangaman commented:
This is a feature I'd love to see added. I've been following the progress in this and related issues and PRs. Regarding the most recent comments, I think it would make sense to leave rescan=true as is, and have rescan=utxo only check an address or key against the utxo set and not do a full rescan for historical transactions.

I'm new to contributing to this project but let me know if I can help with testing or anything for this PR. Or, @ryanofsky, if you don't feel like you'll be revisiting this issue in the foreseeable future maybe I can try to pick up where you left off. Thanks.

ryanofsky commented:
@sangaman, it'd be great if you want to pick this up. As mentioned in my last update I haven't worked on this on a while and don't have plans to continue at the moment.
